### PR TITLE
Internationalization

### PR DESCRIPTION
**Pull request by bmcdorman**
This PR adds internationalization support to the simulator (i.e., all strings in the UI are now translatable). See the README for documentation on how this system works.

# Screenshots

## Changing Locale
image

## After Change
image

### CODE REVIEW

File: i18n/i18n.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{
```
Comment (navzam) [inline i18n/i18n.json:1]: Is `i18n.json` supposed to be checked in?

File: src/Sim.tsx [hunk 2]
```diff
@@ -35,6 +36,9 @@ import ScriptManager from './ScriptManager';
 import Geometry from './state/State/Scene/Geometry';
 import Camera from './state/State/Scene/Camera';
 
+import tr from '@i18n';
+
+const TEST = tr('test');
```
Comment (navzam) [inline src/Sim.tsx:41]: Remove

File: src/components/SettingsDialog.tsx [hunk 3]
```diff
@@ -76,11 +90,23 @@ interface SectionProps {
   selected?: boolean;
 }
 
-export class SettingsDialog extends React.PureComponent<Props, State> {
+const LOCALE_OPTIONS: ComboBox.Option[] = (() => {
+  const ret: ComboBox.Option[] = [];
+  for (const locale of [LocalizedString.EN_US, LocalizedString.JA_JP]) {
```
Comment (navzam) [inline src/components/SettingsDialog.tsx:95]: Remove `JA_JP` before merge?

File: package.json [hunk 4]
```diff
@@ -12,18 +12,22 @@
     "test": "env TS_NODE_PROJECT=\"tsconfig.testing.json\" mocha --recursive --extension ts --require ts-node/register test",
     "start": "yarn run start-dev",
     "start-dev": "webpack-dev-server --host 0.0.0.0 --config=configs/webpack/dev.js",
-    "start-prod": "yarn run build && node express.js"
+    "start-prod": "yarn run build && node express.js",
+    "generate-i18n": "ts-node i18n/generate.ts",
+    "build-i18n": "ts-node i18n/build.ts"
```
Comment (navzam) [inline package.json:17]: If `i18n.json` isn't checked in, does `yarn run build-i18n` need to be added as a step in CD?

File: README.md [hunk 5]
```diff
@@ -119,6 +119,29 @@ The project is set up with [ESLint](https://eslint.org/) for JavaScript/TypeScri
 
 To ease development, we highly recommend enabling ESLint within your editor so you can see issues in real time. If you're using Visual Studio Code, you can use the [VS Code ESLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint). For other editors, see [available ESLint integrations](https://eslint.org/docs/user-guide/integrations).
 
+## Internationalization (i18n)
+
+Simulator leverages gettext PO files to create a `i18n.json` file located in `/i18n/i18n.json`. The source files are scanned for imports of `@i18n`, and uses of the default exported function (`tr` by convention) are detected and inserted into PO files located in `/i18n/po/`. These PO files are "built" into the JSON file, which is then injected into the frontend via webpack's `DefinePlugin` (see `configs/webpack/common.js` for details). The `tr` function reads this object at runtime to make translations available.
+
+To update the PO files, run `yarn run generate-i18n`. While the generation script *should* preserve your work-in-progress, it is recommended to commit, stash, or otherwise backup the PO files prior to running this script.
```
Comment (navzam) [inline README.md:126]: What's the expectation for devs regarding PO files? Should `generate-i18n` be run before every PR?

For example, if I add a new UI string, should I regenerate the PO files and include the modified PO files in my PR, even if I don't fill in the translations?

File: src/util/LocalizedString.ts [hunk 6]
```diff
@@ -72,6 +72,41 @@ namespace LocalizedString {
...
+    typeof HE_IL |
+    typeof TH_TH |
+    typeof VI_VN |
+    typeof ID_ID |
+    typeof MS_MY |
+    typeof FA_IR |
+    typeof UR_PK
+  );
+
   export const FALLBACKS: { [locale: string]: string[] } = {
```
Comment (navzam) [inline src/util/LocalizedString.ts:110]: Should the fallbacks (other than `EN_US`) be transitive? For example, currently `ES_MX` can fallback to `ES_ES` but not vice-versa

File: configs/webpack/common.js [hunk 7]
```diff
@@ -131,6 +141,7 @@ module.exports = {
       SIMULATOR_HAS_CPYTHON: JSON.stringify(dependencies.cpython !== undefined),
       SIMULATOR_HAS_AMMO: JSON.stringify(dependencies.ammo !== undefined),
       SIMULATOR_LIBKIPR_C_DOCUMENTATION: JSON.stringify(libkiprCDocumentation),
+      SIMULATOR_I18N: JSON.stringify(i18n),
```
Comment (navzam) [inline configs/webpack/common.js:144]: Any way to include `i18.n.json` in a way that it gets picked up by `yarn watch`?

File: src/components/SimMenu.tsx [hunk 8]
```diff
@@ -255,7 +267,7 @@ class SimMenu extends React.PureComponent<Props, State> {
           disabled={false}
         >
           <ItemIcon icon={faStop} />
-          Stop
+          {LocalizedString.lookup(tr('Stop', 'Terminate program execution'), locale)}
```
Comment (navzam) [inline src/components/SimMenu.tsx:270]: Doesn't get translated, even though the translation exists in the PO and JSON files. Seems to be a pattern where none of the `tr(str, str)` (with the second param) work properly

File: src/components/SimMenu.tsx [hunk 9]
```diff
@@ -265,7 +277,7 @@ class SimMenu extends React.PureComponent<Props, State> {
           style={{ borderLeft: `1px solid ${theme.borderColor}` }}
         >
           <ItemIcon icon={faPlay} />
-          Run
+          {LocalizedString.lookup(tr('Run', 'Begin program execution'), locale)}
         </RunItem>
       );
```
Comment (navzam) [inline src/components/SimMenu.tsx:283]: github won't let me comment on unchanged lines, but "Layout" and "World" below need `tr()`

Comment (navzam) [reply src/components/SimMenu.tsx:283]: and "Reset World"

File: src/components/ChallengeRoot.tsx [hunk 10]
```diff
@@ -10,12 +10,12 @@ import { styled } from 'styletron-react';
...
+import AboutDialog from './AboutDialog';
 
-import { FeedbackDialog } from './Feedback';
+import FeedbackDialog from './Feedback';
 import { sendFeedback, FeedbackResponse } from './Feedback/SendFeedback';
-import { FeedbackSuccessDialog } from './Feedback/SuccessModal';
+import FeedbackSuccessDialog from './Feedback/SuccessModal';
 
 import compile from '../compile';
 import { SimulatorState } from './SimulatorState';
```
Comment (navzam) [inline src/components/ChallengeRoot.tsx:21]: "Welcome to the KIPR Simulator!" needs `tr()`

File: src/components/Root.tsx [hunk 11]
```diff
@@ -10,12 +10,12 @@ import { styled } from 'styletron-react';
...
+import AboutDialog from './AboutDialog';
 
-import { FeedbackDialog } from './Feedback';
+import FeedbackDialog from './Feedback';
 import { sendFeedback, FeedbackResponse } from './Feedback/SendFeedback';
-import { FeedbackSuccessDialog } from './Feedback/SuccessModal';
+import FeedbackSuccessDialog from './Feedback/SuccessModal';
 
 import compile from '../compile';
 import { SimulatorState } from './SimulatorState';
```
Comment (navzam) [inline src/components/Root.tsx:21]: "Welcome to the KIPR Simulator!" needs `tr()`

File: src/components/ChallengeRoot.tsx (hunk 10 above)
Comment (navzam) [reply src/components/ChallengeRoot.tsx:21]: and "Compiling..." and "Compilation succeeded" and "Compilation failed" etc.

File: src/components/Root.tsx (hunk 11 above)
Comment (navzam) [reply src/components/Root.tsx:21]: and "Compiling..." and "Compilation succeeded" and "Compilation failed" etc.

File: src/components/Info/Info.tsx [hunk 12]
```diff
@@ -18,18 +18,25 @@ import Motor from '../../AbstractRobot/Motor';
 import { faSync } from '@fortawesome/free-solid-svg-icons';
```
Comment (navzam) [inline src/components/Info/Info.tsx:18]: In `componets/Info/Location.tsx`, these need `tr()`:
- X
- Y
- Z
- Rotation

File: src/components/Info/Info.tsx [hunk 13]
```diff
@@ -94,14 +101,6 @@ const MOTOR_POSITIONS_NAME = StyledText.compose({
...
 
-const ANALOG_NAME = StyledText.text({
-  text: 'Analog Sensors',
-});
-
-const DIGITAL_NAME = StyledText.text({
-  text: 'Digital Sensors',
-});
-
 const ResetIcon = styled(Fa, ({ theme }: ThemeProps) => ({
```
Comment (navzam) [inline src/components/Info/Info.tsx:104]: `tr()` needed for:
- Servos
- Motor Velocities
- Motor Positions

File: src/components/World/index.tsx [hunk 14]
```diff
@@ -441,7 +447,7 @@ class World extends React.PureComponent<Props & ReduxWorldProps, State> {
     const itemsName = StyledText.compose({
```
Comment (navzam) [inline src/components/World/index.tsx:447]: Item names themselves aren't being translated, possible due to hardcoded `EN_US` below:

```
itemList.push(EditableList.Item.standard({
        component: Item,
        props: { name: node.name[LocalizedString.EN_US], theme },
```

File: src/components/Feedback/index.tsx [hunk 15]
```diff
@@ -46,7 +56,7 @@ export const FeedbackLink = styled('a', () => ({
   color: 'lightblue',
 }));
 
-export class FeedbackDialog extends React.PureComponent<Props, State> {
+class FeedbackDialog extends React.PureComponent<Props, State> {
   constructor(props: Props) {
     super(props);
     this.state = {
```
Comment (navzam) [inline src/components/Feedback/index.tsx:62]: Needs `tr()` below: "Give a helpful description of..."

File: src/components/AboutDialog.tsx [hunk 16]
```diff
@@ -10,11 +10,23 @@ import { faCopyright } from '@fortawesome/free-solid-svg-icons';
 import KIPR_LOGO_BLACK from '../assets/KIPR-Logo-Black-Text-Clear-Large.png';
```
Comment (navzam) [inline src/components/AboutDialog.tsx:10]: (ignore location of this comment)

Documentation window didn't get any `tr()`s. Specifically seeing "Documentation" window title and "Search..." in search box

File: README.md (hunk 5 above)
Comment (bmcdorman) [reply README.md:126]: `generate-i18n` probably should be run before PRs, but the result is the same; If no translations exist the fallback is always `EN_US`, which is specified directly in the source code.

File: src/util/LocalizedString.ts (hunk 6 above)
Comment (bmcdorman) [reply src/util/LocalizedString.ts:110]: I don't know enough to answer that question. Think we'd have to ask a translator what the best action is. I think we should leave as-is for now.

File: configs/webpack/common.js (hunk 7 above)
Comment (bmcdorman) [reply configs/webpack/common.js:144]: We can watch for it, but not sure how to specify "as a part of the watch, re-read the JSON file".

File: src/components/OpenSceneDialog.tsx [hunk 17]
```diff
@@ -113,7 +116,7 @@ class OpenSceneDialog extends React.PureComponent<Props & ReduxOpenSceneDialogPr
     });
 
     return (
-      <Dialog name='Open World' theme={theme} onClose={onClose}>
+      <Dialog name={LocalizedString.lookup(tr('Open World'), locale)} theme={theme} onClose={onClose}>
         <Container theme={theme}>
           <SceneColumn theme={theme}>
             {loadedScenesArray.map(s => this.createSceneName(s[0], s[1]))}
```
Comment (navzam) [inline src/components/OpenSceneDialog.tsx:122]: "Accept" button below needs `tr()`

File: src/components/World/NodeSettings.tsx [hunk 18]
```diff
@@ -966,16 +879,114 @@ class NodeSettings extends React.PureComponent<Props, State> {
...
+      ComboBox.option(LocalizedString.lookup(tr('Axis Angle'), locale), 'angle-axis'),
+    ];
+    
+    const EULER_ORDER_OPTIONS: ComboBox.Option[] = [
+      ComboBox.option('XYZ', 'xyz'),
+      ComboBox.option('YZX', 'yzx'),
+      ComboBox.option('ZXY', 'zxy'),
+      ComboBox.option('XZY', 'xzy'),
+      ComboBox.option('YXZ', 'yxz'),
+      ComboBox.option('ZYX', 'zyx'),
```
Comment (navzam) [inline src/components/World/NodeSettings.tsx:920]: needs `tr()`... I think? I'm not sure how it'd get translated, but we have 'X' 'Y' and 'Z' translated elsewhere to might as well

**Comment by navzam** [comment]
Probably out of scope for this PR, but I created an issue for localization of the login/signup page: https://github.com/kipr/Simulator/issues/459

File: configs/webpack/common.js (hunk 7 above)
Comment (navzam) [reply configs/webpack/common.js:147]: Hm yeah. I think it'd be nice to have when working on translations. We can create an issue for now

File: src/util/LocalizedString.ts (hunk 6 above)
Comment (navzam) [reply src/util/LocalizedString.ts]: Works for me

File: README.md (hunk 5 above)
Comment (navzam) [reply README.md:126]: Makes sense. I worry about how quickly our "% of UI translated" will drop considering how rapidly our UI changes. May need to come up with some periodic update with translators

File: README.md [hunk 19]
```diff
@@ -119,6 +119,29 @@ The project is set up with [ESLint](https://eslint.org/) for JavaScript/TypeScri
 
 To ease development, we highly recommend enabling ESLint within your editor so you can see issues in real time. If you're using Visual Studio Code, you can use the [VS Code ESLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint). For other editors, see [available ESLint integrations](https://eslint.org/docs/user-guide/integrations).
 
+## Internationalization (i18n)
```
Comment (navzam) [inline README.md:122]: Best to add `yarn run build-i18n` in the README instructions above since it's a required step now

Comment (navzam) [reply README.md:122]: In hindsight, maybe it's unnecessary for webpack to require `i18n.json` to exist (assuming everything defaults properly without it) *for development*, but I think it's important to ensure it exists for prod builds

File: src/components/OpenSceneDialog.tsx [hunk 20]
```diff
@@ -156,7 +159,7 @@ class OpenSceneDialog extends React.PureComponent<Props & ReduxOpenSceneDialogPr
   };
```
Comment (navzam) [inline src/components/OpenSceneDialog.tsx:159]: scene names are translated but use hardcoded `EN_US` above

File: src/scenes/jbc15b.ts [hunk 21]
```diff
@@ -41,8 +43,8 @@ const REAM2_ORIGIN: ReferenceFrame = {
 
 export const JBC_15B: Scene = {
   ...baseScene,
-  name: { [LocalizedString.EN_US]: 'JBC 15B' },
-  description: { [LocalizedString.EN_US]: `Junior Botball Challenge 15B: Bump Bump` },
+  name: tr('JBC 15B'),
+  description: tr('Junior Botball Challenge 15B: Bump Bump'),
```
Comment (navzam) [inline src/scenes/jbc15b.ts:47]: "Paper Ream 1" and "Paper Ream 2" below need `tr()`

File: src/components/Challenge/EventViewer.tsx [hunk 22]
```diff
@@ -26,17 +26,18 @@ const Description = styled('div', {
...
-        <Name>{LocalizedString.lookup(name, LocalizedString.EN_US)}</Name>
+        <Name>{LocalizedString.lookup(name, locale)}</Name>
         <Spacer />
         {eventState !== undefined && <FontAwesomeIcon icon={eventState ? faCheck : faCross} />}
       </NameContainer>
-      {description && <Description>{LocalizedString.lookup(description, LocalizedString.EN_US)}</Description>}
+      {description && <Description>{LocalizedString.lookup(description, locale)}</Description>}
     </Container>
   );
 };
```
Comment (navzam) [inline src/components/Challenge/EventViewer.tsx:43]: `EventSettings.tsx` has a few hardcoded `EN_US` for event names/descriptions. not sure how those are supposed to be handled

Comment (bmcdorman) [reply src/components/Challenge/EventViewer.tsx:43]: EventSettings isn't currently used and isn't finished. I suggest we leave as is.

**Review by navzam — APPROVED** [review]

**Pull request merged**